bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/HeapSizeManager.java

```diff
@@ -64,60 +74,79 @@ public long getMaxHeapSize() {
   }
 
   public synchronized void waitUntilAllOperationsAreDone() throws InterruptedException {
+    isFlushing.set(true);
     boolean performedWarning = false;
-    while(!pendingOperationsWithSize.isEmpty()) {
+    while (true) {
       cleanupFinishedOperations();
       if (pendingOperationsWithSize.isEmpty()) {
         break;
       }
-      if (!performedWarning
-          && lastOperationChange + INTERVAL_NO_SUCCESS_WARNING < System.currentTimeMillis()) {
-        long lastUpdated = (System.currentTimeMillis() - lastOperationChange) / 1000;
-        LOG.warn("No operations completed within the last %d seconds."
-            + "There are still %d operations in progress.", lastUpdated,
-          pendingOperationsWithSize.size());
+      if (!performedWarning && requiresWarning()) {
+        warn();
         performedWarning = true;
-        waitForCompletions(FINISH_WAIT_MILLIS);
       }
+      wait(FINISH_WAIT_MILLIS);
     }
     if (performedWarning) {
       LOG.info("flush() completed");
     }
+    isFlushing.set(false);
+  }
+
+  private boolean requiresWarning() {
+    return lastOperationChange + INTERVAL_NO_SUCCESS_WARNING < System.currentTimeMillis();
   }
 
+  private void warn() {
+    long lastUpdated = (System.currentTimeMillis() - lastOperationChange) / 1000;
+    LOG.warn(
+        "No operations completed within the last %d seconds."
+        + "There are still %d operations in progress.",
+        lastUpdated, pendingOperationsWithSize.size());
+  }
 
   private void cleanupFinishedOperations() {
-    List<Long> toClean = new ArrayList<>();
-    completedOperationIds.drainTo(toClean);
-    if (!toClean.isEmpty()) {
-      markOperationsCompleted(toClean);
+    if (!completedOperationIds.isEmpty()) {
+      List<Long> toClean = new ArrayList<>();
+      boolean cleaned = false;
+      if (completedOperationIds.drainTo(toClean) > 0) {
+        cleaned = true;
+        for (Long operationSequenceId : toClean) {
+          markOperationComplete(operationSequenceId);
+        }
+        toClean.clear();
+      }
+      if (cleaned) {
+        lastOperationChange = System.currentTimeMillis();
+      }
     }
   }
+
+  private void markOperationComplete(Long operationSequenceId) {
+    Long heapSize = pendingOperationsWithSize.remove(operationSequenceId);
+    if (heapSize != null) {
+      currentWriteBufferSize -= heapSize;
+    } else {
+      LOG.warn("An operation completion was recieved multiple times. Your operations completed."
```

Comment:
The operation was not completed multiple times.  Maybe "An operation was reported to be complete multiple times.  It's likely due to a communcation glitch and a retry."